### ISSUE TITLE
Classifier test function

### ISSUE THREAD

**Issue by lehtonenp** [gtkfi/ArcSDM#231]
# ArcPy-Based Classifier Testing Function for Raster Inputs

This document outlines the process for implementing a regressor testing function using **ArcPy** and **Scikit-learn**. The goal is to test a trained machine learning model on raster inputs and optionally compute evaluation metrics like **MSE**.

Reference Implementation from EIS Toolkit
For inspiration and comparison, see:
[classifier_test_cli](https://github.com/GispoCoding/eis_toolkit/blob/master/eis_toolkit/cli.py#L2801) in the EIS Toolkit repository.

---

## Function Purpose

The test function:

- Reads raster input features.
- Loads a trained regression model (`.joblib`).
- Makes predictions.
- Optionally compares predictions to target labels using metrics.
- Saves predictions as raster output.

---

## Dependencies

```
import arcpy
from arcpy.sa import *
import numpy as np
import joblib
from sklearn.metrics import mean_squared_error
```

Enable the Spatial Analyst extension:
```
arcpy.CheckOutExtension("Spatial")
```
# Step-by-Step Developer Instructions
## Step 1: Read Input Rasters
Use ArcPy to load each input raster.

```
input_rasters = [arcpy.Raster(path) for path in input_paths]
```
Ensure all rasters are spatially aligned.

Use the same projection, extent, and resolution.

## Step 2: Convert Rasters to NumPy Arrays
Use RasterToNumPyArray and stack the arrays into a feature matrix.
```
arrays = [arcpy.RasterToNumPyArray(r, nodata_to_value=np.nan) for r in input_rasters]
X = np.stack(arrays, axis=-1).reshape(-1, len(arrays))
```
Handles NoData as np.nan for compatibility.

Each pixel becomes a feature vector.

## Step 3: Load Trained Model
Load the machine learning model using joblib.

```
model = joblib.load(model_file)
```
Ensure the model was trained on the same input feature format.

## Step 4: Perform Prediction
Use the loaded model to generate predictions.

```
predictions = model.predict(X)
```
Returns a flat array of predictions.

## Step 5: Reshape Predictions to Raster Shape
Convert the 1D predictions back to 2D using raster dimensions.

```
rows, cols = input_rasters[0].height, input_rasters[0].width
prediction_2d = predictions.reshape(rows, cols)
```
Optionally mask out NoData regions.

## Step 6: Score Predictions (Optional)
If a ground truth (target labels) raster is provided:

```
y_true = arcpy.RasterToNumPyArray(arcpy.Raster(target_labels), nodata_to_value=np.nan).flatten()

valid_mask = ~np.isnan(y_true)
mse = mean_squared_error(y_true[valid_mask], predictions[valid_mask])

metrics_dict = {"mse": round(mse, 3)}
```
Add additional metrics as needed (e.g., r2_score, mean_absolute_error).

## Step 7: Save Output Raster
Convert the predicted 2D array into a raster and save it.

```
prediction_raster = arcpy.NumPyArrayToRaster(
    prediction_2d,
    lower_left_corner=input_rasters[0].extent.lowerLeft,
    x_cell_size=input_rasters[0].meanCellWidth,
    y_cell_size=input_rasters[0].meanCellHeight,
    value_to_nodata=np.nan
)

prediction_raster.save(output_raster)
```
Inherit spatial reference from an input raster.

Replace np.nan with an appropriate NoData value if needed.

## Step 8: Logging & Output
Log the output and optionally print evaluation metrics.

```
arcpy.AddMessage("Regression test completed.")
if target_labels:
    arcpy.AddMessage("Metrics:", metrics_dict)
```
Use `arcpy.AddMessage()`, `arcpy.AddWarning()` and `arcpy.AddError()` in ArcGIS for GUI-based tools.

## Notes
Confirm the model's expected input format matches the structure of raster data.

For ArcGIS Pro users: make sure the Spatial Analyst extension is available and licensed.
